mqtt_client.c

```diff
@@ -576,81 +580,88 @@ static esp_err_t mqtt_process_receive(esp_mqtt_client_handle_t client)
-            if (is_valid_mqtt_msg(client, MQTT_MSG_TYPE_SUBSCRIBE, msg_id)) {
-                ESP_LOGD(TAG, "Subscribe successful");
-                client->event.event_id = MQTT_EVENT_SUBSCRIBED;
-                esp_mqtt_dispatch_event(client);
-            }
-            break;
-        case MQTT_MSG_TYPE_UNSUBACK:
-            if (is_valid_mqtt_msg(client, MQTT_MSG_TYPE_UNSUBSCRIBE, msg_id)) {
-                ESP_LOGD(TAG, "UnSubscribe successful");
-                client->event.event_id = MQTT_EVENT_UNSUBSCRIBED;
-                esp_mqtt_dispatch_event(client);
-            }
-            break;
-        case MQTT_MSG_TYPE_PUBLISH:
-            if (msg_qos == 1) {
-                client->mqtt_state.outbound_message = mqtt_msg_puback(&client->mqtt_state.mqtt_connection, msg_id);
-            }
-            else if (msg_qos == 2) {
-                client->mqtt_state.outbound_message = mqtt_msg_pubrec(&client->mqtt_state.mqtt_connection, msg_id);
-            }
-
-            if (msg_qos == 1 || msg_qos == 2) {
-                ESP_LOGD(TAG, "Queue response QoS: %d", msg_qos);
-
-                if (mqtt_write_data(client) != ESP_OK) {
-                    ESP_LOGE(TAG, "Error write qos msg repsonse, qos = %d", msg_qos);
-                    // TODO: Shoule reconnect?
-                    // return ESP_FAIL;
+        msg_type = mqtt_get_type(buffer);
+        msg_qos = mqtt_get_qos(buffer);
+        msg_id = mqtt_get_id(buffer, bytes_to_parse);
+        parsed_length = mqtt_get_total_length(buffer, bytes_to_parse);
+
+        ESP_LOGD(TAG, "msg_type=%d, msg_id=%d", msg_type, msg_id);
+        switch (msg_type)
+        {
+            case MQTT_MSG_TYPE_SUBACK:
+                if (is_valid_mqtt_msg(client, MQTT_MSG_TYPE_SUBSCRIBE, msg_id)) {
+                    ESP_LOGD(TAG, "Subscribe successful");
+                    client->event.event_id = MQTT_EVENT_SUBSCRIBED;
+                    esp_mqtt_dispatch_event(client);
+                }
+                break;
+            case MQTT_MSG_TYPE_UNSUBACK:
+                if (is_valid_mqtt_msg(client, MQTT_MSG_TYPE_UNSUBSCRIBE, msg_id)) {
+                    ESP_LOGD(TAG, "UnSubscribe successful");
+                    client->event.event_id = MQTT_EVENT_UNSUBSCRIBED;
+                    esp_mqtt_dispatch_event(client);
+                }
+                break;
+            case MQTT_MSG_TYPE_PUBLISH:
+                if (msg_qos == 1) {
+                    client->mqtt_state.outbound_message = mqtt_msg_puback(&client->mqtt_state.mqtt_connection, msg_id);
+                }
+                else if (msg_qos == 2) {
+                    client->mqtt_state.outbound_message = mqtt_msg_pubrec(&client->mqtt_state.mqtt_connection, msg_id);
                 }
-            }
-            client->mqtt_state.message_length_read = read_len;
-            client->mqtt_state.message_length = mqtt_get_total_length(client->mqtt_state.in_buffer, client->mqtt_state.message_length_read);
```

Comment:
As far as I can see you've removed support for msg total length. This would not work for mqtt messages with len> MQTT Buffer Size (from menuconfig)